File: 2024/Day 3 - Mull It Over/Day3.py
import re

with open('data3.txt') as file:
    data = file.read()

sum = 0
x = re.findall("mul[(]\d+,\d+[)]", data)

for i in x:
    y = re.findall(r"\d+", i)
    sum += int(y[0])*int(y[1])

print(sum)

#................Part 2..................

sum2 = 0
flag = True
x2 = re.findall(r"mul[(]\d+,\d+[)]|do[(][)]|don't[(][)]", data)

for k in x2:
    if k == "do()":
        flag = True
    elif k == "don't()":
        flag = False
    else:
        if flag:
            c = re.findall(r"\d+", k)
            sum2 += int(c[0])*int(c[1])

print(sum2)

File: 2024/Day 3 - Mull It Over/Day3.cpp
#include<iostream>
#include<fstream>
#include<regex>
using namespace std;

int computer(string line)
{
    int result = 0;
    regex search = regex(R"(mul\((\d+),(\d+)\))");
    
    auto words_begin = sregex_iterator(line.begin(), line.end(), search);
    auto words_end = sregex_iterator();
    
    for(sregex_iterator i = words_begin; i != words_end; i++)
    {
        smatch m = *i;
        result += stoi(m[1].str())*stoi(m[2].str());
    }
    return result;
}

int conditionalComputer(string line)
{
    int result = 0;
    bool flag = true;
    regex search = regex(R"(mul\((\d+),(\d+)\)|do\(\)|don't\(\))");
    
    auto words_begin = sregex_iterator(line.begin(), line.end(), search);
    auto words_end = sregex_iterator();
    
    for(sregex_iterator i = words_begin; i != words_end; i++)
    {
        smatch m = *i;
        if(m[0].str() == "do()")
            flag = true;
        else if (m[0].str() == "don't()")
            flag = false;
        else
            if(flag)
                result += stoi(m[1].str())*stoi(m[2].str());
    }
    return result;
}

int main()
{
    ifstream datafile("data3.txt");
    string line, data;
    while (getline(datafile, line))
        data += line;
    
    cout<<computer(data)<<endl;
    cout<<conditionalComputer(data)<<endl;
}
